### PR TITLE
Add vector value linking

### LINKED ISSUE #144: Add a link/unlink option for subproperties like "scale"
**Describe the project you are working on:**
2D action game, card game

**Describe the problem or limitation you are having in your project:**
Throughout the day I change properties like `scale` in the editor many, many times. Which already can be quite annoying because of https://github.com/godotengine/godot/issues/28618.
In 95% of the cases I want to change properties like `scale` proportionally. So I have to first set the x, then the y value, _both to the same value_. Because the editor does not offer to link those properties, all those operations have to be done twice ... every ... single ... time.

**Describe how this feature / enhancement will help you overcome this problem or limitation:**
A small clickable link/unlink icon next to values that would often be set proportionally. 
A small mouseover text explaining what it does, for those unsure about the meaning of the icon, similar to mouseover on properties.
I would expect `scale` to be **linked** by default, others may be unlinked by default.

**Show a mock up screenshots/video or a flow diagram explaining how your proposal will work:**
linked_properties

**Describe implementation detail for your proposal (in code), if possible:**
I'm C++ illiterate.

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**
It's so common everywhere else, people are used to that level of comfort. I expect it to be used a lot.

**Is there a reason why this should be core and not an add-on in the asset library?:**
Noone will/would/should download addons for such small UI quality of life improvements.

### PR DESCRIPTION
**Pull request by KoBeWi**
Closes https://github.com/godotengine/godot-proposals/issues/144

Linked values change proportionally (unless any of them is 0). The link button only appears if property has `PROPERTY_HINT_LINK` (for now only Node2D/3D/Control `scale`).

The linked/unlinked icon is a placeholder, I did my best :/ I'll optimize them when they are replaced.

On a side note, I'm thinking about adding possibility to link 2 properties. This would be mostly useful for min/max values in ParticlesMaterial. So I wonder if the new property hint could be reused for that in the future.

### CODE REVIEW

**Comment by fire-forge** [comment]
This looks great! However, I think the link button should be moved to the right side as it'd make more sense there than on the left.

Also, are you considering doing this for Vector3 as well? The same argument about scale usually being the same on all axes can be applied to 3D too.

**Comment by KoBeWi** [comment]
> Also, are you considering doing this for Vector3 as well?

Not in this PR. I didn't want it too big and we still need to agree on the design of this feature.

**Comment by akien-mga** [comment]
> On a side note, I'm thinking about adding possibility to link 2 properties. This would be mostly useful for min/max values in ParticlesMaterial. So I wonder if the new property hint could be reused for that in the future.

This is actually possible already, see `ADD_LINKED_PROPERTY` in `object.h`.

**Comment by groud** [comment]
As discussed on a the PR meeting, the PR is very good but would need small changes:
- The link button should be hidden when the hint is not here. Later on, if we find out the there are situations where the link would be useful but off by default, we could add another hint to specify the default value. But for now there is no need to have them for all Vector2.
- I think it might depend on how it looks like, but removing the button will make the x/y components not aligned from one property to another. As a consequence it would be interesting to try moving it the right, to see if it looks better.

As a side note (that was not discussed in the PR meeting), do you think you could make the same changes for Vector3/Vector3i too?

**Comment by KoBeWi** [comment]
> This is actually possible already, see ADD_LINKED_PROPERTY in object.h.

This is for dependent properties, it doesn't allow for conditional linking (like with the link button).

> As a side note (that was not discussed in the PR meeting), do you think you could make the same changes for Vector3/Vector3i too?

Yes, I was already planning to do it, just waited for the feature approval before putting more work.

**Comment by KoBeWi** [comment]
Ok, reworked a bit. Moved the icon to the right and made the button hidden without the link hint:
image
I'll now add it to Vector3.

**Comment by KoBeWi** [comment]
Ok added new icon contributed by @redlamp:

The linked icon he provided was the same as Instance icon, so I reused Instance icon. It could be renamed to Linked maybe, but not in this PR.

File: editor/editor_properties.cpp [hunk 1]
```diff
@@ -1801,6 +1862,27 @@ void EditorPropertyVector3::_value_changed(double val, const String &p_name) {
...
+		ratio_xz = spin[0]->get_value() / spin[2]->get_value();
+		ratio_yz = spin[1]->get_value() / spin[2]->get_value();
+	} else {
+		ratio_yx = 1.0;
+		ratio_zx = 1.0;
+		ratio_xy = 1.0;
+		ratio_zy = 1.0;
+		ratio_xz = 1.0;
+		ratio_yz = 1.0;
+	}
```
Comment (akien-mga) [inline editor/editor_properties.cpp:1885]: `ratio_yx` is the same as `1.0 / ratio_xy`.
Maybe you could reduce the amount of member variables and computation needed by keeping only 3 and using `*` or `/` operations as appropriate?

Comment (KoBeWi) [reply editor/editor_properties.cpp:1885]: I could, I even tried it at first, but IIRC there were some precision issues.

**Comment by fire-forge** [comment]
> On a side note, I'm thinking about adding possibility to link 2 properties. This would be mostly useful for min/max values in ParticlesMaterial. So I wonder if the new property hint could be reused for that in the future.

If you do add this in the future, perhaps it could be used to link the corner radius and border width properties of StyleBoxFlat. I find that I usually use the same values for all 4 sides/corners and having the link button there would make it easier to edit them all at once.

image

**Review by akien-mga — APPROVED** [review]
Should be good to merge once rebased.

**Comment by akien-mga** [comment]
Thanks!

**Comment by nathanfranke** [comment]
> > On a side note, I'm thinking about adding possibility to link 2 properties. This would be mostly useful for min/max values in ParticlesMaterial. So I wonder if the new property hint could be reused for that in the future.
> 
> If you do add this in the future, perhaps it could be used to link the corner radius and border width properties of StyleBoxFlat. I find that I usually use the same values for all 4 sides/corners and having the link button there would make it easier to edit them all at once.
> 
> image

Maybe these should be a single property with type [`Vector4`](https://github.com/godotengine/godot-proposals/issues/629) with a new property hint like `PROPERTY_HINT_SIDES` that changes `x`/`y`/`z`/`w` to `left`/`right`/`up`/`down` (or whatever the best order is).

**Comment by timothyqiu** [comment]
Cherry-picked for 3.6

**Pull request merged**